merry-go-round.md

```diff
@@ -0,0 +1,133 @@
+component of the Portal Network to deliver an efficient mechanism for syncing
+the Ethereum state.
+
+## Storage Network
+
+We build off of the current preliminary design of the state network, focusing
+on the following elements.
+
+### 1. Preservation of locality
+
+All leaf data is mapped onto the DHT in a manner that preserves locality.  This
+means that nodes store contiguous sections of trie data and two nodes that are
+*close* to each other in the DHT will store similar data.
+
+> Note that this matters primarily for the main account trie, and for large
+> contract storage tries.  For tries that do not contain very much data, that
+> data will be spread very thinly around the DHT and thus nodes are unlikely to
+> store more than a single leaf node of data.
+
+### 2. Keeping data up-to-date with gossiped proofs
+
+At each block, the state network's gossip mechanism will disseminate proofs at
+each new state root that allow individual nodes to update their state data from
+previous state roots to newer state roots.  These updates are done as merkle
+proofs to ensure they can be verified against the newer state root and don't
+require trust.
+
+## Basic Design of the Merry-Go-Round Sync
+
+### Naive/Simplistic Sync
+
+The high level design of the "merry go round" is that a node wishing to sync
+the state would iterate through all of the nodes in the network, requesting the
+proofs they are storing about the state, and merging them into a single
+wholistic view of the entire trie.  In an environment where the Ethereum state
+was static, and didn't change, this approach would result in having a full
+copy of all of the state data after a node had fully traversed the DHT and
+recieved all of the individual state proofs.
+
+The Ethereum state is however, constantly changing as each new block is added
+to the chain.  The result is that as a node walks through the network
+requesting state data, the first proofs they receive will be anchored to some
+state root that we'll call `S_n` and at some point, the proofs they receive will
+be anchored to some subsequent root that we'll call `S_n+1`.
+
+### Dealing with updates to the state
+
+* Problem: As the Ethereum state changes, nodes that are syncing the state must
+  be able to *update* the data they have already received to newer and newer
+  state roots.
+
+To address this, we will introduce two changes to how nodes in the network
+handle state.
+
+First, we introduce a concept of epoch boundaries.  Nodes in the network will
+store proof of the segment of the state that they are storing for the network
+at the most recent state root from the most recent epoch boundary.  The result
+of this is that within a single epoch, all requests for state data from
+different nodes will be anchored to the same state root.
+
+Under this model we still assume that all updates to the state are being
+gossiped as each block is added to the chain.  Nodes that are storing data
```

Comment:
How realistic is this assumption? What is the probability for some nodes to keep lagging behind new block state updates and not catch up with all state updates until the end of the "epoch"?

Reply:
I think that some nodes will lag behind by a few blocks but that this won't have any significant negative impact on the system.